### PR TITLE
Feature/targets maximum

### PR DESCRIPTION
**Pull request by t-romani**
the limit of targets per user is 10
after creating 10 targets the user should't be able to create another target, and receive an error.

### CODE REVIEW

File: app/models/target.rb [hunk 1]
```diff
@@ -32,4 +32,12 @@ class Target < ApplicationRecord
   validates :title, presence: true, uniqueness: {
     scope: %i[topic_id user_id], case_sensitive: false
   }
+
+  before_save :limit_number_of_targets
```
Comment (gonzaM) [inline app/models/target.rb]: Can't you use a validation?

Comment (t-romani) [reply app/models/target.rb]: Couldn't come up with anything, any ideas or useful links @gonzaM ?

Comment (gonzaM) [reply app/models/target.rb]: Do you know custom validations? https://guides.rubyonrails.org/active_record_validations.html#custom-methods
Something like:
```ruby
validate :limit_number_of_targets

def limit_number_of_targets
  return if user has less than 10
  errors.add(:base, "A custom error...")
end
```

File: app/errors/TargetLimitReached.rb [hunk 2]
```diff
@@ -0,0 +1 @@
+class TargetLimitReached < StandardError; end
```
Comment (JuanfraM) [inline app/errors/TargetLimitReached.rb]: if you call this file `target_limit_reached_exception.rb` then you need to import it in the api controller?

File: spec/request/target/create_spec.rb [hunk 3]
```diff
@@ -71,5 +70,24 @@
         end
       end
     end
+
+    context 'when limit reached' do
+      before do
+        10.times do
+          user.targets.create!(attributes_for(:target, topic_id: topic.id))
+        end
```
Comment (JuanfraM) [inline spec/request/target/create_spec.rb]: use `create_list` for this purpose from [FactoryBot](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md)

File: app/models/target.rb (hunk 1 above)
Comment (JuanfraM) [reply app/models/target.rb]: I think what @gonzaM said is more appropriate in this case. This validation only apply in the target model

**Review by JuanfraM — COMMENTED** [review]
Nice job!
I left some comments

File: spec/request/target/create_spec.rb [hunk 4]
```diff
@@ -71,5 +70,24 @@
...
+      end
```
Comment (andresg4) [inline spec/request/target/create_spec.rb:78]: Maybe here you can use the `create_list` from `FactoryBot` that lets you create multiple records ([create_list](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#building-or-creating-multiple-records)).
You could do something like this:
```ruby
before do
  create_list(:target, 10, user: user, topic: topic)
  subject
end
```
And with this you can also avoid the `subject` in each test.

File: app/errors/TargetLimitReached.rb (hunk 2 above)
Comment (t-romani) [reply app/errors/TargetLimitReached.rb]: this is no longer in the proyect

File: app/models/target.rb (hunk 1 above)
Comment (t-romani) [reply app/models/target.rb]: this is creating me some issues with the creation of targets when the user is null, for example on belong_to spec validation.

File: app/models/target.rb [hunk 5]
```diff
@@ -32,4 +32,15 @@ class Target < ApplicationRecord
   validates :title, presence: true, uniqueness: {
     scope: %i[topic_id user_id], case_sensitive: false
   }
+
+  validate :limit_number_of_targets
+
+  private
+
+  def limit_number_of_targets
+    return if user && user.targets.count < 10
```
Comment (gonzaM) [inline app/models/target.rb:41]: <=?

**Review by gonzaM — APPROVED** [review]

File: spec/request/target/create_spec.rb [hunk 6]
```diff
@@ -71,5 +70,21 @@
...
       end
     end
+
+    context 'when limit reached' do
+      before do
+        create_list(:target, 10, user: user, topic: topic)
+        subject
+      end
+
+      it 'gets an unauthorized response' do
```
Comment (andresg4) [inline spec/request/target/create_spec.rb]: Maybe `'gets a bad request response'`?

**Review by andresg4 — APPROVED** [review]
👏

File: app/models/target.rb (hunk 5 above)
Comment (t-romani) [reply app/models/target.rb:41]: I think < works, because the limit is exactly 10

File: spec/request/target/create_spec.rb (hunk 6 above)
Comment (t-romani) [reply spec/request/target/create_spec.rb]: 🤦‍♀

**Review by JuanfraM — APPROVED** [review]

File: app/controllers/api/v1/targets_controller.rb [hunk 7]
```diff
@@ -16,6 +16,10 @@ def destroy
         targets.find(params[:id]).destroy!
       end
 
+      def destroy
+        user.targets.find(params[:id]).destroy!
```
Comment (JuanfraM) [inline app/controllers/api/v1/targets_controller.rb]: You've already declared the helper `targets`, you can use it here.
I think it was added in the PR of delete targets but it seems it was overridden here.

**Review by JuanfraM — APPROVED** [review]
Just one small comment

**Pull request merged**